### PR TITLE
WeightedFairScheduler: double or nothing

### PR DESCRIPTION
**Pull request by levaitamas**
WeightedFairScheduler is not able to keep up with shares set in 'count' aka batch domain. 

_tl;dr: The wf scheduler is inaccurate with non-extreme shares. Representing internal values as doubles solves this inaccuracy._ 

As an example, take a simple tree pipeline with a single 'root' node and two 'leaf' nodes. These are  connected by a RandomSplit. Root and leaf modules are put into separate tasks (see config at the bottom). Suppose, we want to schedule in batch domain ('count') and share ratios (leaf:root) are set as 1:1.71 and 1:1.7. To achieve that, shares are set to 100 (leaf), and 171 or 170 (root).

We expect to see a small difference in measured latency in the two scenarios. Actually, we see a large performance gap:
```
$ ./bessctl run local/wfq-test root_share=1.7
FLOW0:
   avg_ns: 1064.000
  90th_ns: 1600.000
  95th_ns: 2200.000
  99th_ns: 3200.000

FLOW1:
   avg_ns: 1063.000
  90th_ns: 1600.000
  95th_ns: 2200.000
  99th_ns: 3200.000

$ ./bessctl run local/wfq-test root_share=1.71
FLOW0:
   avg_ns: 42947.000
  90th_ns: 71700.000
  95th_ns: 74500.000
  99th_ns: 100700.000

FLOW1:
   avg_ns: 35842.000
  90th_ns: 69900.000
  95th_ns: 73200.000
  99th_ns: 96200.000
```

To solve this issue, the important code section is in [WeightedFairTrafficClass::FinishAndAccountTowardsRoot()](https://github.com/NetSys/bess/blob/master/core/traffic_class.cc#L278); where pass_delta is calculated. The pass_delta is important for the underlying stride scheduling, since the scheduler takes the task with the smallest pass value at each iteration.

If we calculate pass_delta values by hand, we see a small difference between pass_delta values corresponding to share 170 and 171:

```
| share | pass_delta |
|-------+------------+
|   100 |    10.2400 |
|   170 |     6.0235 |
|   171 |     5.9883 |
```

But, pass_delta is represented as an integer and BESS is writen in C++, so pass_delta will be 6 for share 170 and 5 for 171. That is a game changing difference in this situation, the ratio of pass_delta values in the later case is not 1.71:1, but 2:1!

Floating-point representation of weighted-fair scheduler internal values solved the performance issue:
```
$ ./bessctl run local/wfq-test root_share=1.7
FLOW0:
   avg_ns: 1060.000
  90th_ns: 1600.000
  95th_ns: 2300.000
  99th_ns: 3100.000

FLOW1:
   avg_ns: 1063.000
  90th_ns: 1600.000
  95th_ns: 2300.000
  99th_ns: 3100.000

$ ./bessctl run local/wfq-test root_share=1.71
FLOW0:
   avg_ns: 1086.000
  90th_ns: 1600.000
  95th_ns: 2300.000
  99th_ns: 3100.000

FLOW1:
   avg_ns: 1087.000
  90th_ns: 1600.000
  95th_ns: 2400.000
  99th_ns: 3100.000
```

WDYT?

---

A BESS script to reproduce results:
```python
import time

root_share  = float($root_share!'1.7')

shares = {'leaf': 100,
          'root': int(root_share * 100)}

src = Source()
rr = RandomSplit(gates=[0,1], drop_rate=0.0)
q1 = Queue()
q2 = Queue()
m1 = Measure()
m2 = Measure()

src -> Timestamp() -> rr
rr:0 -> q1 -> m1 -> Sink()
rr:1 -> q2 -> m2 -> Sink()

bess.add_tc('root', policy='weighted_fair', resource='count')
src.attach_task(parent='root', share=shares['root'])
q1.attach_task(parent='root', share=shares['leaf'])
q2.attach_task(parent='root', share=shares['leaf'])

bess.resume_all()

time.sleep(2)

bess.pause_all()

for i,m in enumerate((m1,m2)):
    percentiles = [90, 95, 99]
    res = m.get_summary(latency_percentiles=percentiles)
    print('FLOW%d:' % i)
    print('   avg_ns: %.3f' % res.latency.avg_ns)
    for i,p in enumerate(percentiles):
        print('  %dth_ns: %.3f' % (
            p,
            res.latency.percentile_values_ns[i]))
    print()
```

### CODE REVIEW

**Comment by barath** [comment]
Nice catch.  Could you run `traffic_class_bench` (suitably modified, if needed) to see what difference the switch to `double` might have made to scheduler overhead?

**Comment by levaitamas** [comment]
Thanks! Means of 10 measurements on the [dev VM](https://github.com/NetSys/bess/blob/master/env/Vagrantfile) running on server CPUs are the following.

### uint64_t
```
-----------------------------------------------------------------------------------------
Benchmark                                                  Time           CPU Iterations
-----------------------------------------------------------------------------------------
TCWeightedFair/TCScheduleOnceCount/4/0_mean               63 ns         63 ns   11118153
TCWeightedFair/TCScheduleOnceCount/8/0_mean               66 ns         66 ns   10630055
TCWeightedFair/TCScheduleOnceCount/16/0_mean              69 ns         69 ns   10214932
TCWeightedFair/TCScheduleOnceCount/32/0_mean              72 ns         72 ns    9555621
TCWeightedFair/TCScheduleOnceCount/64/0_mean              78 ns         78 ns    8824669
TCWeightedFair/TCScheduleOnceCount/128/0_mean             90 ns         90 ns    7897144
TCWeightedFair/TCScheduleOnceCount/256/0_mean            109 ns        109 ns    6551838
TCWeightedFair/TCScheduleOnceCount/512/0_mean            130 ns        130 ns    5297849
TCWeightedFair/TCScheduleOnceCount/1024/0_mean           159 ns        159 ns    4365361
TCWeightedFair/TCScheduleOnceCount/2048/0_mean           190 ns        190 ns    3724089
TCWeightedFair/TCScheduleOnceCount/4096/0_mean           202 ns        202 ns    3529551
TCWeightedFair/TCScheduleOnceCount/8192/0_mean           206 ns        206 ns    3405542
TCWeightedFair/TCScheduleOnceCount/16384/0_mean          215 ns        215 ns    3295705
TCWeightedFair/TCScheduleOnceCount/32768/0_mean          290 ns        290 ns    2348607
TCWeightedFair/TCScheduleOnceCount/65536/0_mean          405 ns        404 ns    1750586
TCWeightedFair_BigO                                    17.88 lgN      17.86 lgN
TCWeightedFair_RMS                                        25 %         25 %
TCWeightedFair/TCScheduleOnceCycle/4/1_mean               66 ns         66 ns   10537643
TCWeightedFair/TCScheduleOnceCycle/8/1_mean               68 ns         68 ns   10249466
TCWeightedFair/TCScheduleOnceCycle/16/1_mean              76 ns         76 ns    9261521
TCWeightedFair/TCScheduleOnceCycle/32/1_mean              85 ns         84 ns    8282332
TCWeightedFair/TCScheduleOnceCycle/64/1_mean              97 ns         97 ns    7214514
TCWeightedFair/TCScheduleOnceCycle/128/1_mean            114 ns        114 ns    6146597
TCWeightedFair/TCScheduleOnceCycle/256/1_mean            133 ns        133 ns    5171720
TCWeightedFair/TCScheduleOnceCycle/512/1_mean            152 ns        152 ns    4680950
TCWeightedFair/TCScheduleOnceCycle/1024/1_mean           173 ns        173 ns    4145995
TCWeightedFair/TCScheduleOnceCycle/2048/1_mean           213 ns        213 ns    3215277
TCWeightedFair/TCScheduleOnceCycle/4096/1_mean           247 ns        247 ns    2787794
TCWeightedFair/TCScheduleOnceCycle/8192/1_mean           270 ns        270 ns    2532794
TCWeightedFair/TCScheduleOnceCycle/16384/1_mean          287 ns        286 ns    2443331
TCWeightedFair/TCScheduleOnceCycle/32768/1_mean          368 ns        367 ns    1916091
TCWeightedFair/TCScheduleOnceCycle/65536/1_mean          596 ns        522 ns    1000000
TCWeightedFair_BigO                                    23.07 lgN      22.26 lgN
TCWeightedFair_RMS                                        46 %         28 %
```

### double
```
----------------------------------------------------------------------------------------
Benchmark                                                  Time           CPU Iterations
----------------------------------------------------------------------------------------
TCWeightedFair/TCScheduleOnceCount/4/0_mean               63 ns         63 ns   11126742
TCWeightedFair/TCScheduleOnceCount/8/0_mean               66 ns         66 ns   10660545
TCWeightedFair/TCScheduleOnceCount/16/0_mean              68 ns         68 ns   10238280
TCWeightedFair/TCScheduleOnceCount/32/0_mean              74 ns         74 ns    9747640
TCWeightedFair/TCScheduleOnceCount/64/0_mean              79 ns         79 ns    8977111
TCWeightedFair/TCScheduleOnceCount/128/0_mean             93 ns         93 ns    7653888
TCWeightedFair/TCScheduleOnceCount/256/0_mean            119 ns        119 ns    5547388
TCWeightedFair/TCScheduleOnceCount/512/0_mean            141 ns        141 ns    4637043
TCWeightedFair/TCScheduleOnceCount/1024/0_mean           168 ns        168 ns    3997264
TCWeightedFair/TCScheduleOnceCount/2048/0_mean           192 ns        192 ns    3586653
TCWeightedFair/TCScheduleOnceCount/4096/0_mean           211 ns        211 ns    3389813
TCWeightedFair/TCScheduleOnceCount/8192/0_mean           218 ns        217 ns    3349284
TCWeightedFair/TCScheduleOnceCount/16384/0_mean          221 ns        221 ns    3203967
TCWeightedFair/TCScheduleOnceCount/32768/0_mean          297 ns        297 ns    2387625
TCWeightedFair/TCScheduleOnceCount/65536/0_mean          413 ns        412 ns    1764310
TCWeightedFair_BigO                                    18.48 lgN      18.47 lgN
TCWeightedFair_RMS                                        24 %         24 %
TCWeightedFair/TCScheduleOnceCycle/4/1_mean               67 ns         67 ns   10399510
TCWeightedFair/TCScheduleOnceCycle/8/1_mean               70 ns         70 ns    9854923
TCWeightedFair/TCScheduleOnceCycle/16/1_mean              78 ns         78 ns    9168463
TCWeightedFair/TCScheduleOnceCycle/32/1_mean              87 ns         87 ns    8099838
TCWeightedFair/TCScheduleOnceCycle/64/1_mean              99 ns         99 ns    7123087
TCWeightedFair/TCScheduleOnceCycle/128/1_mean            120 ns        120 ns    5270220
TCWeightedFair/TCScheduleOnceCycle/256/1_mean            138 ns        138 ns    5078838
TCWeightedFair/TCScheduleOnceCycle/512/1_mean            164 ns        163 ns    4398370
TCWeightedFair/TCScheduleOnceCycle/1024/1_mean           189 ns        189 ns    3567482
TCWeightedFair/TCScheduleOnceCycle/2048/1_mean           227 ns        227 ns    3125388
TCWeightedFair/TCScheduleOnceCycle/4096/1_mean           257 ns        257 ns    2681737
TCWeightedFair/TCScheduleOnceCycle/8192/1_mean           281 ns        280 ns    2499371
TCWeightedFair/TCScheduleOnceCycle/16384/1_mean          297 ns        297 ns    2370191
TCWeightedFair/TCScheduleOnceCycle/32768/1_mean          381 ns        381 ns    1813489
TCWeightedFair/TCScheduleOnceCycle/65536/1_mean          566 ns        562 ns    1000000
TCWeightedFair_BigO                                    23.51 lgN      23.46 lgN
TCWeightedFair_RMS                                        31 %         30 %
```

**Comment by levaitamas** [comment]
@sangjinhan and @chris3torek could you please take a look at this PR?

**Comment by melvinw** [comment]
Thanks! cat
